File: reboot/examples/bank-zod/api/bank/v1/account.ts
import { reader, writer } from "@reboot-dev/reboot-api";
import { z } from "zod/v4";

export const Account = {
  state: {
    balance: z.number().default(0).meta({ tag: 1 }),
  },
  methods: {
    balance: reader({
      request: {},
      response: {
        amount: z.number().meta({ tag: 1 }),
      },
    }),
    deposit: writer({
      request: {
        amount: z.number().meta({ tag: 1 }),
      },
      response: z.void(),
    }),
    withdraw: writer({
      request: {
        amount: z.number().meta({ tag: 1 }),
      },
      response: z.void(),
      errors: [
        z
          .object({
            type: z.literal("OverdraftError"),
            amount: z.number().meta({ tag: 1 }),
          })
          .meta({ tag: 1 }),
      ] as const,
    }),
    open: writer({
      factory: {},
      request: {},
      response: z.void(),
    }),
    interest: writer({
      request: {},
      response: z.void(),
    }),
  },
};

export const api = {
  Account,
};

File: reboot/examples/bank-pydantic/api/bank/v1/pydantic/account.py
from reboot.api import API, Field, Methods, Model, Reader, Type, Writer


class AccountState(Model):
    balance: float = Field(tag=1)


class BalanceResponse(Model):
    amount: float = Field(tag=1)


class DepositRequest(Model):
    amount: float = Field(tag=1)


class WithdrawRequest(Model):
    amount: float = Field(tag=1)


class OverdraftError(Model):
    amount: float = Field(tag=1)


AccountMethods = Methods(
    balance=Reader(
        request=None,
        response=BalanceResponse,
        mcp=None,
    ),
    deposit=Writer(
        request=DepositRequest,
        response=None,
        mcp=None,
    ),
    withdraw=Writer(
        request=WithdrawRequest,
        response=None,
        errors=[OverdraftError],
        mcp=None,
    ),
    # Must use this method to create an instance of Account.
    open=Writer(
        request=None,
        response=None,
        factory=True,
        mcp=None,
    ),
    interest=Writer(
        request=None,
        response=None,
        mcp=None,
    ),
)

api = API(
    Account=Type(
        state=AccountState,
        methods=AccountMethods,
    ),
)
